spark/src/main/scala/com/gu/tableversions/spark/VersionedDataset.scala

```diff
@@ -115,24 +116,19 @@ object VersionedDataset {
   /**
     * Write the given partitioned dataset, storing each partition in the associated path.
     */
-  private[spark] def writeVersionedPartitions[T](dataset: Dataset[T], partitionPaths: Map[Partition, URI]): Unit = {
-    // This is a slow and inefficient implementation that writes each partition in sequence,
-    // we can look into a more performant solution later.
-
-    def filteredForPartition(partition: Partition): Dataset[T] =
-      partition.columnValues.foldLeft(dataset) {
-        case (filteredDataset, partitionColumn) =>
-          filteredDataset.where(s"${partitionColumn.column.name} = '${partitionColumn.value}'")
-      }
+  private[spark] def writeVersionedPartitions[T](
+      dataset: Dataset[T],
+      table: TableDefinition,
+      partitionVersions: Map[Partition, Version])(implicit spark: SparkSession): Unit = {
 
-    val datasetsWithPaths: Map[Dataset[T], URI] =
-      partitionPaths.map { case (partition, path) => filteredForPartition(partition) -> path }
+    VersionedFileSystem.writeConfig(VersionedFileSystemConfig(partitionVersions),
+                                    dataset.sparkSession.sparkContext.hadoopConfiguration)
 
-    datasetsWithPaths.foreach {
-      case (datasetForPartition, partitionPath) =>
-        datasetForPartition.write
-          .parquet(partitionPath.toString) // TODO: Take in format parameter. Or a dataset writer?
-    }
-  }
+    val partitions = table.partitionSchema.columns.map(_.name)
 
+    dataset.write
+      .partitionBy(partitions: _*)
+      .mode(SaveMode.Append)
+      .parquet(VersionedFileSystem.scheme + "://" + table.location.getPath) // TODO: Take in format parameter. Or a dataset writer?
+  }
```

Comment:
This is really the key change of this PR: instead of writing partitions manually, we get Spark to do it for us, the versioning of partitions being handled by the underlying filesystem proxy.